### PR TITLE
API-40893-slack-alert-single-instance

### PR DESCRIPTION
**Pull request by stiehlrod**
## Summary

- Adds check for the last day, compares transaction_ids and only pulls in new Transaction_ids from the last hour.

## Related issue(s)

- [API-40893](https://jira.devops.va.gov/browse/API-40893)

## Testing done

- [X] Used testing channel for slack alerts
[Instructions reference](https://github.com/department-of-veterans-affairs/vets-api/pull/15648)
1. Trigger alerts by adding the following lines to a controller
```
require 'claims_api/report_hourly_unsuccessful_submissions'
ClaimsApi::ReportHourlyUnsuccessfulSubmissions.new.perform
```
2. Add `audit_enabled: true` to settings.local
Screenshot 2024-09-25 at 2 43 18 PM

3. Need to have the slack channel_id, and webhook_url in settings.local
see [#vaapi-team-dash-engineers](https://oddball.slack.com/archives/C03LPB6HJR5/p1727371828133059?thread_ts=1727295330.949129&cid=C03LPB6HJR5) for this info since it is considered a secret

4. Need to flip on `claims_hourly_slack_error_report_enabled`
5. Need to have sidekiq running: `bundle exec sidekiq` - `http://localhost:3000/sidekiq/`

6. Alert to test channel
Screenshot 2024-09-26 at 11 46 53 AM

7. I created 5 claims in my local db: created_at within the hour, cid: '0oagdm49ygCSJTp8X297', 3 different transactions_ids. the alert should group by transaction_id, and grab one claim_id from each group and send that array tot he channel.

## Screenshots
_Note: Optional_

## What areas of the site does it impact?
	modified:   modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb
	modified:   modules/claims_api/spec/sidekiq/report_hourly_unsuccessful_submissions_spec.rb

## Acceptance criteria

- [X]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [X]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

### CODE REVIEW

File: modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb [hunk 1]
```diff
@@ -59,7 +59,7 @@ def notify(claims, va_claims, poa, itf, ews, from, to, env)
 
     def errored_submissions_exist?
       [@errored_claims, @va_gov_errored_claims, @errored_poa, @errored_itf, @errored_ews].any? do |var|
```
Comment (FonzMP) [inline modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb:61]: don't know if we need the internal logic here. any would report if any exist. count positive is essentially duplicating that concept.

File: modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb [hunk 2]
```diff
@@ -68,13 +68,25 @@ def allow_processing?
...
-                                              )
-      va_gov.group_by(&:transaction_id)
+      errored_claims = get_unique_errors
+      errored_claims&.group_by(&:transaction_id)
+    end
+
+    def get_unique_errors
+      starting = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+
+      two_days = ClaimsApi::AutoEstablishedClaim.where(created_at: 48.hours.ago..@search_to,
```
Comment (FonzMP) [inline modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb:78]: I think we may be mixing concepts here. The original thought was to query the last 24 hours as that's the retry window VA.gov uses, once we do that, compare the past 30 min with the 23 1/2 hours prior to and compare the uniques, so same concept, but we have a 48 hour window defined here. 

I'm assuming the concept of the "2 days" may be a mixup of us discussing the possibility of trying to use the nightly job, which wouldn't work either way. This is an assumption as to why the 48 hour window is above, as the ticket defined 24 hours

File: modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb [hunk 3]
```diff
@@ -8,15 +8,15 @@ class ReportHourlyUnsuccessfulSubmissions < ClaimsApi::ServiceBase
     def perform
       return unless allow_processing?
 
-      @search_to = 1.minute.ago
+      @search_to = Time.zone.now
```
Comment (FonzMP) [inline modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb:11]: I remember a ticket that made some attempt of not falsely alerting as we thought there may have been some issue with submissions coming in at the same time as the hourly report, and somehow falsely reporting errored status to our Slack alert - is the change from 1.minute.ago to now a needed fix? 

we've also modified it [here](https://github.com/department-of-veterans-affairs/vets-api/pull/18606/files) - perhaps this is needed for this query to work, just want to make sure we're not re-introducing something we intentionally changed?

File: modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb [hunk 4]
```diff
@@ -68,13 +68,25 @@ def allow_processing?
...
+    end
+
+    def get_unique_errors
+      starting = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+
+      one_day = ClaimsApi::AutoEstablishedClaim.where(created_at: 24.hours.ago..@search_to,
+                                                      status: 'errored', cid: '0oagdm49ygCSJTp8X297')
+
+      unique = one_day.uniq { |claim| claim[:transaction_id] } if one_day.present?
+      errored_claims = unique.select { |claim| claim[:created_at] >= @search_from } if unique.present?
```
Comment (tycol7) [inline modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb:82]: I _think_ the idea is to get the errored claims from the past 30 minutes and remove any that appear in the `24 hours ago` _up to_ `30 minutes ago` range (by transaction id), the logic being that we must have already reported on them.

As written, we will report all unique errored claims in the past hour, which could result in duplicate reported claims on successive runs.

**Review by tycol7 — COMMENTED** [review]
Looking good, but I think we will still report duplicate claims as written.

File: modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb [hunk 5]
```diff
@@ -68,32 +67,32 @@ def allow_processing?
...
-      @va_gov_errored_claims = [] if @va_gov_errored_claims.blank?
-      @va_gov_errored_claims << errored_claims
-      if @va_gov_errored_claims.flatten.blank?
-        {}
-      else
-        @va_gov_errored_claims.flatten&.group_by(&:transaction_id)
-        @va_gov_errored_claims&.map { |grp| grp[1][0] }&.pluck(:id)
-      end
+      errored_claims = get_unique_errors
+      errored_claims.map { |record| record[0] } # rubocop:disable Rails/Pluck
```
Comment (tycol7) [inline modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb:71]: Why not  `get_unique_errors.pluck(:id)`? Or better yet, return `errored_claims.pluck(:id)` from `get_unique_errors` and remove this method? 🤷

File: modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb [hunk 6]
```diff
@@ -68,32 +67,32 @@ def allow_processing?
...
-      errored_claims = unique.select { |claim| claim[:created_at] >= @search_from } if unique.present?
-      # Elapsd time: 0.01658499985933304 seconds
+        day_trans_ids = last_day.pluck(:transaction_id)
 
-      ending = Process.clock_gettime(Process::CLOCK_MONOTONIC)
-      elapsed = ending - starting
-      ClaimsApi::Logger.log('hourly_slack_alert_elapsed_time', detail: "Elapsd time: #{elapsed} seconds")
+        errored_claims = last_hour.find_all do |claim|
+          day_trans_ids.exclude?(claim[:transaction_id])
+        end
```
Comment (tycol7) [inline modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb]: 👍

also unlikely, but if there are no errors in `last_day`, we will hit an error on line 84.

File: modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb [hunk 7]
```diff
@@ -68,32 +67,32 @@ def allow_processing?
...
+      rescue ActiveRecord::Exception => e
+        ClaimsApi::Logger "SQL error in #{e}"
+        ActiveRecord::Base.connection.execute 'ROLLBACK'
 
-      errored_claims
+        raise e
+      end
```
Comment (tycol7) [inline modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb]: I _think_ we should be ok without this rescue block. These queries donʼt seem likely to fail, and there isnʼt anything to rollback if they do.

File: modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb [hunk 8]
```diff
@@ -68,32 +67,32 @@ def allow_processing?
...
+      errored_claims&.pluck(:id, :transaction_id)
```
Comment (tycol7) [inline modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb:95]: could we just return `id` here? or are we using `transaction_id` somewhere?

File: modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb [hunk 9]
```diff
@@ -68,13 +67,32 @@ def allow_processing?
...
+        day_trans_ids = last_day.pluck(:transaction_id)
+
+        errored_claims = last_hour.find_all do |claim|
+          day_trans_ids.exclude?(claim[:transaction_id])
+        end
+      rescue ActiveRecord::Exception => e
+        ClaimsApi::Logger "SQL error in #{e}"
+        ActiveRecord::Base.connection.execute 'ROLLBACK'
+
+        raise e
```
Comment (tycol7) [inline modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb:93]: I _think_ we should be ok without this rescue block. The queries are straightforward and there is nothing to rollback.

File: modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb [hunk 10]
```diff
@@ -68,13 +67,32 @@ def allow_processing?
...
+
+        last_hour = ClaimsApi::AutoEstablishedClaim
+                    .where(created_at: 1.hour.ago..Time.zone.now,
+                           status: 'errored', cid: '0oagdm49ygCSJTp8X297')
+
+        day_trans_ids = last_day.pluck(:transaction_id)
+
+        errored_claims = last_hour.find_all do |claim|
+          day_trans_ids.exclude?(claim[:transaction_id])
+        end
```
Comment (tycol7) [inline modules/claims_api/app/sidekiq/claims_api/report_hourly_unsuccessful_submissions.rb:88]: 👍

also unlikely, but if there are no errors in `last_day`, we will hit an error on line 84.

**Review by tycol7 — DISMISSED** [review]
lgtm!

**Comment by FonzMP** [comment]
good catch on the timeframe. I thought this job ran 30 minutes when we wrote this ticket but it's obviously hourly. I'll note on the ticket the correction.

**Pull request merged**